experiments/compositions/composition/internal/controller/expander_reconciler.go

```diff
@@ -329,65 +332,24 @@ func (r *ExpanderReconciler) SetupWithManager(mgr ctrl.Manager, cr *unstructured
 		Complete(r)
 }
 
-func (r *ExpanderReconciler) getExpanderVersion(ctx context.Context, inputExpanderVersion string, expanderType string) (string, string, error) {
+func (r *ExpanderReconciler) getExpanderVersion(ctx context.Context, inputExpanderVersion string, expanderType string) (string, *compositionv1alpha1.ExpanderVersion, error) {
 	logger := log.FromContext(ctx)
 
-	var expanderVersionCR compositionv1alpha1.ExpanderVersion
-	var expanderVersion, expanderRegistry string
-	var availableVersions []string
-	err := r.Client.Get(ctx, types.NamespacedName{Name: "composition-" + expanderType, Namespace: "composition-system"}, &expanderVersionCR)
+	var ev compositionv1alpha1.ExpanderVersion
+	err := r.Client.Get(ctx, types.NamespacedName{Name: "composition-" + expanderType, Namespace: "composition-system"}, &ev)
 	if err != nil {
 		// The CR should be created before the specified expander can be used.
-		return "", "", err
-	}
-	expanderRegistry = expanderVersionCR.Spec.ImageRegistry
-	availableVersions = expanderVersionCR.Spec.ValidVersions
-	SemVerVersions := make([]*semver.Version, len(availableVersions))
-	for i, r := range availableVersions {
-		v, err := semver.NewVersion(r)
-		if err != nil {
-			logger.Info("Error parsing version: %s", err)
-		}
-		SemVerVersions[i] = v
-	}
-	sort.Sort(semver.Collection(SemVerVersions))
-	logger.Info("inputexpanderverions", "current", inputExpanderVersion)
-	// Currently the version will default to latest if not set.
-	if inputExpanderVersion != "latest" {
-		// Verify if the input version is valid.
-		semVersion, err := semver.NewVersion(inputExpanderVersion)
-		if err != nil {
-			return "", "", fmt.Errorf("invalidExpanderVersion")
-		}
-		if !isValidVersion(SemVerVersions, semVersion) {
-			return "", "", fmt.Errorf("invalidExpanderVersion")
-		}
-		expanderVersion = inputExpanderVersion
-	} else {
-		// The existing semver package sort only supports pure numbers.
-		expanderVersion = "v" + SemVerVersions[len(availableVersions)-1].String() + ".alpha"
+		return "", nil, err
 	}
-	return expanderVersion, expanderRegistry, nil
-}
 
-func isValidVersion(availableVersions []*semver.Version, version *semver.Version) bool {
-	var low, high, mid int
-	low = 0
-	high = len(availableVersions) - 1
-	for low <= high {
-		mid = low + (high-low)/2
+	if ev.Status.VersionMap == nil {
+		return "", nil, fmt.Errorf("ExpanderVersion .status.versionMap is empty")
```

Comment:
It seems like you are suggesting to use versionMap for version validation instead of the validversions spec field. What do we do if the versionMap is empty? I don't see an error handling in expander_reconciler for this error.

Reply:
this is generated from the validversions. this is the status field that is populated by the expanderversion reconciler. 
https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/experiments/compositions/composition/internal/controller/expanderversion_controller.go#L111